File: src/main/kotlin/themoviedb/TheMovieDbClient.kt
package themoviedb

import org.funktionale.either.Either.Left
import org.funktionale.either.Either.Right
import services.HttpService
import utils.CS
import utils.Serie
import utils.TimeRange
import utils.Utils.warn
import java.time.LocalDate
import java.time.format.DateTimeParseException

object TheMovieDbClient {

    private val apiKey = "000ffc8b6e767158ff5489a8daba11c2"
    private val baseUrl = "https://api.themoviedb.org/3"


    fun getBestSeriesAtPage(page: Int = 1): CS<List<Serie>> =
        HttpService.httpGetThrottled(
            "$baseUrl/discover/tv",
            DiscoverEndpoint.Result::class,
            "api_key" to apiKey,
            "sort_by" to "popularity.desc",
            "page" to page.toString()
        )
            .thenApply {
                when (it) {
                    is Right ->
                        it.right().get().results
                            .filterNotNull()
                            .map { Serie(it.id, it.name) }
                    is Left -> {
                        warn("NotFound when fetching series at page $page")
                        emptyList()
                    }
                }
            }

    fun getSeasonsNumbers(serie: Serie): CS<List<Int>> =
        HttpService.httpGetThrottled(
            "$baseUrl/tv/${serie.id}",
            TvShowEndpoint.TvShow::class,
            "api_key" to apiKey
        )
            .thenApply {
                when (it) {
                    is Right ->
                        it.right().get().seasons
                            .map { it.season_number }
                            .filterNot { it == 0 }
                    is Left -> {
                        warn("NotFound when fetching season numbers of serie $serie")
                        emptyList()
                    }
                }
            }.exceptionally {
                warn("Failed to parse the season numbers of tvShow $serie")
                emptyList()
            }

    fun getSeasonTimeRange(serie: Serie, season: Int): CS<TimeRange?> =
        HttpService.httpGetThrottled(
            "$baseUrl/tv/${serie.id}/season/$season",
            SeasonEndpoint.Season::class,
            "api_key" to apiKey
        )
            .thenApply {
                when (it) {
                    is Right ->
                        it.right().get().episodes
                            .map { it.air_date }
                            .filterNotNull()
                            .map {
                                try {
                                    LocalDate.parse(it)
                                } catch (e: DateTimeParseException) {
                                    null
                                }
                            }
                            .filterNotNull()
                            .sorted()
                    is Left -> {
                        warn("NotFound when fetching season $season of serie $serie")
                        emptyList()
                    }
                }
            }.thenApply {
                if (it.isEmpty()) {
                    warn("${serie.name} season $season has zero episodes")
                    null
                } else {
                    TimeRange(it.first(), it.last())
                }
            }
}

File: src/themoviedb/themoviedbClient.ts
import { reverse } from 'esrever'
import axios from 'axios'
import { Serie, TimeRange } from '../myTypes'

const API_KEY = reverse('2c11abad8a9845ff851767e6b8cff000')
const BASE_URL = 'https://api.themoviedb.org/3'

type DiscoverEndpointResult = {
  results: { id: number; name: string }[]
}
type TvShowEndpointResult = {
  seasons: { season_number: number }[]
}
type SeasonEndpointResult = {
  episodes: { air_date: string }[]
}

// TODO gérer tous les cas foireux que je gérais en Kotlin pour être fault tolerant

export async function getBestSeriesAtPage({ page = 1 } = {}): Promise<Serie[]> {
  const {
    data: { results },
  } = await axios.get<DiscoverEndpointResult>(`${BASE_URL}/discover/tv`, {
    params: {
      api_key: API_KEY,
      sort_by: 'popularity.desc',
      page,
    },
  })
  return results
}

export async function getSeasonsNumbers(serie: Serie): Promise<number[]> {
  const { data } = await axios.get<TvShowEndpointResult>(
    `${BASE_URL}/tv/${serie.id}`,
    {
      params: {
        api_key: API_KEY,
        append_to_response: 'seasdons',
      },
    },
  )
  return data.seasons.map((_) => _.season_number)
}

export async function getSeasonTimeRange(
  serie: Serie,
  season: number,
): Promise<TimeRange> {
  const {
    data: { episodes },
  } = await axios.get<SeasonEndpointResult>(
    `${BASE_URL}/tv/${serie.id}/season/${season}`,
    {
      params: {
        api_key: API_KEY,
      },
    },
  )

  const [start, end] = [episodes[0], episodes[episodes.length - 1]].map(
    ({ air_date }) => {
      if (!air_date.match(/^\d{4}-\d{2}-\d{2}$/)) {
        throw new Error(
          `Unexpected episode date :  ${serie.id}, ${serie.name}, ${season}, ${air_date}`,
        )
      }
      return new Date(`${air_date}T00:00:00Z`)
    },
  )
  return { start, end }
}
